lib/libxdp/tests/test_xsk_refcnt.c

```diff
@@ -0,0 +1,308 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+
+#include <errno.h>
+#include <linux/err.h>
+#include <net/if.h>
+#include <pthread.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/resource.h>
+#include <unistd.h>
+
+#include "logging.h"
+
+#include <xdp/libxdp.h>
+#include <xdp/xsk.h>
+
+typedef __u64 u64;
+typedef __u32 u32;
+typedef __u16 u16;
+typedef __u8  u8;
+
+#define MAX_EVENTS 10
+#define MAX_NUM_QUEUES 4
+
+struct xsk_umem_info {
+	struct xsk_ring_prod fq;
+	struct xsk_ring_cons cq;
+	struct xsk_umem *umem;
+	void *buffer;
+};
+
+struct xsk_socket_info {
+	struct xsk_ring_cons rx;
+	struct xsk_umem_info *umem;
+	struct xsk_socket *xsk;
+};
+
+struct xsk_test_event {
+	u32 num_create;
+	u32 num_delete;
+	u32 create_qids[MAX_NUM_QUEUES];
+	u32 delete_qids[MAX_NUM_QUEUES];
+};
+
+struct xsk_test {
+	char name[64];
+	u32 num_events;
+	struct xsk_test_event events[MAX_EVENTS];
+};
+
+/* Tests that use less queues must come first
+ */
+static struct xsk_test all_tests[] = {
+	{"Single socket created and deleted", 2,
+		{{1, 0, {0}, {}}, {0, 1, {}, {0}}}},
+	{"2 sockets, created and deleted sequentially", 4,
+		{{1, 0, {0}, {}}, {1, 0, {1}, {}}, {0, 1, {}, {0}},
+		 {0, 1, {}, {1}}}},
+	{"2 sockets, created sequentially and deleted asynchronously", 3,
+		{{1, 0, {0}, {}}, {1, 0, {1}, {}}, {0, 2, {}, {0, 1}}}},
+	{"2 sockets, asynchronously delete and create", 3,
+		{{1, 0, {0}, {}}, {1, 1, {1}, {0}}, {0, 1, {}, {1}}}},
+	{"3 sockets, created and deleted sequentially", 6,
+		{{1, 0, {0}, {}}, {1, 0, {1}, {}}, {1, 0, {2}, {}},
+		 {0, 1, {}, {1}}, {0, 1, {}, {2}}, {0, 1, {}, {0}}}},
+};
```

Comment:
While this is nice and compact, it also looks like a really weird lisp dialect ;)

Could we maybe fold out the recursive arrays a bit and add some comments on what they're supposed to be doing?